### ISSUE TITLE
Output variables are declared and stored on every element block 

### ISSUE THREAD

**Issue by jwfoulk** [gahansen/Albany#83]
It seems that all output variables are stored on every element block. This is handled through an element block truth table in Exodus. This is not required and can lead to significantly larger output files if particular models (read crystal plasticity) have an order of magnitude more element variables than other models (read J2). I don't know how to fix it. I just stumbled onto this problem while debugging a failed test. 

The solution would be to have a truth table in exodus that represents a block by block representation of the variables for that particular block. Many thanks.

**Comment by bartgol** [comment]
That's a significant concern also regarding the side-set equations framework (which, to be honest, are still under development and not fully supported). For instance, it would be nice to store unknowns defined on a 2D side-domain just on that mesh, rather than storing them everywhere.

I believe the issue is how the discretizaton class builds the DOF's numberation from the mesh entities. As of now, DOFs are numbered like dof_id = node_id*num_equations + eq_id (or node_id+num_nodes*eq_id, depending on the dof ordering). This would be fine if the eq_id was a "global  equation id", and if every mesh stored only the "global equation ids" of the equations defined on that mesh.

This could be done not just on a mesh basis, but also on a mesh-block basis. When the discretization builds the DOFs, instead of looping on the mesh nodes, we could loop on the element buckets and process the nodes of those elements according to the block they're in (and the equations defined in there). Notice that this would require to keep track of the pairs (node_id, eq_id) already processed, so that we do not create the same dof twice when handling nodes at the interface of two elements. But it would be memory uses only in the setup phase and released immediately, so no big deal (but must be done to avoid errors!).

I think this is deeper than just LCM. I think this may involve a noticeable redesigning of parts of the discretization.

**Comment by lxmota** [comment]
@jwfoulk @bartgol I agree that this is a significant problem, specially in the case that @jwfoulk describes.

I will raise this issue at the next Albany teleconference. Thanks for pointing it out.
